File: python/63_powerful_digit_counts.py
def powerful_digit_counts() -> int:
    
    # if an n-th power x^n has n digits,
    # then it must fulfill the inequality
    # 10^(n - 1) <= x^n < 10^n
    # x^n < 10^n means that x can be at most 9
    # 10^(n - 1) <= x^n becomes false when n becomes too large
    
    powers: int = 0
    
    for x in range(1, 10):
        n = 1
        while x ** n >= 10 ** (n - 1):
            powers += 1
            n += 1
    
    return powers
    
if __name__ == "__main__":
    print(powerful_digit_counts())

File: java/_63_PowerfulDigitCounts.java
import java.util.*;
import java.math.BigInteger;

public class _63_PowerfulDigitCounts{
    
    public static void main(String[] args){
        
        // Find all integers p with n digits that are also nth powers.
        // Rewriting p as x^n, these integers must fulfill the following condition:
        // 10^(n-1) <= x^n < 10^n
        
        // Since 10^0 has 1 digit, 10^1 has 2 digits, 10^2 has 3 digits, ...,
        // fulfilling the condition x^n < 10^n requires x to be at most 9.
        // Only nth powers of x = 1, 2, 3, ..., 9 need to be checked.
        // The nth powers of x are then iterated until 10^(n-1) > x^n.
        
        int numPowers = 0;
        
        for (int x = 1; x < 10; x++){
            for (int n = 1; BigInteger.valueOf(10).pow(n-1).compareTo(BigInteger.valueOf(x).pow(n)) <= 0; n++){
                // System.out.println(String.format("%d^%d = %s", x, n, BigInteger.valueOf(x).pow(n)));
                numPowers++;
            }
        }
        
        System.out.println("n-digit powers that are also nth powers: " + numPowers); // 49
    }
}